Add StrahlkorperInDifferentFrameAligned

markscheel opened the pull request:
## Proposed changes

    This is a simplified version of StrahlkorperInDifferentFrame,
    that assumes:
     * The map preserves the center of the Strahlkorper.
     * For each point on the Strahlkorper, the map preserves the angles
       between that point and the center.

Edit: this *does* know about the distorted frame now.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

markscheel committed:
Add StrahlkorperInDifferentFrameAligned.

This is a simplified version of StrahlkorperInDifferentFrame,
that assumes:
 * The map preserves the center of the Strahlkorper.
 * For each point on the Strahlkorper, the map preserves the angles
   between that point and the center.

knelli2 commented on src/ApparentHorizons/StrahlkorperInDifferentFrameHelpers.hpp:
```diff
@@ -0,0 +1,35 @@
+/// \endcond
+
+namespace strahlkorper_in_different_frame_helpers {
+template <typename SrcFrame, typename DestFrame>
+void coords_to_different_frame(
```
Can you add a small amount of docs? Just saying that this loops over the blocks of the domain and transforms the coords from the source frame to the destination frame.

knelli2 commented on src/ApparentHorizons/StrahlkorperInDifferentFrameAligned.hpp:
```diff
@@ -0,0 +1,53 @@
+/// strahlkorper_in_different_frame to map between arbitrary frames;
+/// allowing strahlkorper_in_different_frame to work on more frames
+/// requires adding member functions to Block.
+template <typename SrcFrame, typename DestFrame>
+void strahlkorper_in_different_frame_aligned(
```
Can you place this in the same file as the original function? These are very related and another file just adds cmake complications.

knelli2 commented on src/ApparentHorizons/StrahlkorperInDifferentFrameAligned.hpp:
```diff
@@ -0,0 +1,53 @@
+///
+/// The destination Strahlkorper is given the same l_max() and m_max()
+/// as the source Strahlkorper.
+///
+/// Note that because the Blocks inside the Domain allow access to
```
Use `\note`

knelli2 reviewed (changes requested):
Just a couple things. Related to one of the comments, if you put both functions in the same file, I think you can just put the helper function in an anonymous namespace at the top of the file and that way you don't add any new files.

markscheel commented:
I put everything in one file, used \note, added some docs, and squashed.

markscheel commented:
Clang-tidy errors appear to all be from our good friend toms748.

markscheel commented:
There was a question of whether I can test this function using the Shape TimeDependence instead of the SphericalCompression TimeDependence.  Currently the answer is no, because the Shape TimeDependence has no way of setting a spherically-symmetric map (other than the identity).  I want a spherically-symmetric map so that I can analytically compute the mapped Strahlkorper independently, for the test.

knelli2 commented:
The map coefficients are stored in the functions of time. Could you just make your own functions of time rather than using the ones from the domain creator?

markscheel commented:
I found a way to use TimeDependence::Shape for the test.  So here's what I did:

* I changed the test to use TimeDependence::Shape
* I closed #4968 and #4969
* I incorporated the distorted-frame stuff in #4969 into this PR (which is not that big of change now that I didn't need #4968 anymore).

knelli2 commented on src/ApparentHorizons/StrahlkorperInDifferentFrame.hpp:
```diff
@@ -26,7 +26,7 @@ class not_null;
 /// source Strahlkorper.
 ///
-/// Note that because the Blocks inside the Domain allow access to
+/// \note Because the Blocks inside the Domain allow access to
 /// maps only between a selected subset of frames, we cannot use
 /// strahlkorper_in_different_frame to map between arbitrary frames;
 /// allowing strahlkorper_in_different_frame to work on more frames
```
Can you add to this note that this function *only* works for SrcFrame = Grid and DestFrame = Inertial.

knelli2 commented on src/ApparentHorizons/StrahlkorperInDifferentFrame.hpp:
```diff
@@ -39,3 +39,25 @@ void strahlkorper_in_different_frame(
+/// invariant. If those assumptions are not satisfied, then
+/// strahlkorper_in_different_frame_aligned will give the wrong answer
+/// and you must use strahlkorper_in_different_frame.  Those
+/// assumptions are satisfied for the shape and size maps but not for
+/// general maps.
```
And then add a note here that this only works for SrcFrame = Grid and DestFrame = Distorted or Inertial. Just to make it clear that one of these functions can be used in the distorted frame (currently) but the other one can't.

knelli2 reviewed (commented):
Looks good. Please squash

markscheel commented:
Squashed (and rebased onto current develop)

knelli2 reviewed (dismissed):
@sxs-collaboration/spectre-core-devs Could one of you look at this? It's needed for size control.

wthrowe commented on src/ApparentHorizons/StrahlkorperInDifferentFrame.cpp:
```diff
@@ -263,10 +293,61 @@ void strahlkorper_in_different_frame(
+  auto& radius = get<::Tags::TempScalar<4>>(temp_buffer);
+
+  StrahlkorperTags::ThetaPhiCompute<SrcFrame>::function(
+      make_not_null(&src_theta_phi), src_strahlkorper);
```
These tag functions are free functions (`StrahlkorperFunctions::theta_phi` here).  Call those instead of going through the tags.

markscheel replied:
Done!

wthrowe commented:
Looks good.  Squash.

markscheel commented:
Squashed.

wthrowe reviewed (approved):
clang-tidy failures are just the usual TOMS stuff.
